File: mobile/androidApp/src/main/kotlin/com/mentora/android/ui/components/AnswerOption.kt
package com.mentora.android.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.unit.dp
import com.mentora.android.theme.MentoraDimens
import com.mentora.android.theme.extendedColors
import com.mentora.android.theme.stateOpacities

/** The 5 states `design-system/COMPONENTS.md` § Answer Options (lines 722-734) defines. */
enum class AnswerOptionState { Default, SelectedUnsubmitted, Correct, Incorrect, DisabledPostSubmit }

/** Test-only hooks (`ui.test.onNodeWithTag`) for asserting the Correct/Incorrect icon actually
 *  renders alongside the text+color signal — mirrors [MentoraFieldErrorIconTestTag]'s rigor. */
const val AnswerOptionCorrectIconTestTag = "mentora-answer-option-correct-icon"
const val AnswerOptionIncorrectIconTestTag = "mentora-answer-option-incorrect-icon"

/**
 * Resolves the 5-state table from plain booleans: not submitted yet -> Default/SelectedUnsubmitted;
 * submitted -> every option matching the real correct answer renders Correct (revealing which one
 * was right, whether or not the user picked it), the user's own wrong pick renders Incorrect, and
 * every other unselected/non-correct option renders DisabledPostSubmit — the standard "reveal the
 * right answer" quiz-result pattern the spec's state names imply (`Correct (after submit)` /
 * `Incorrect (after submit)` are keyed to `isCorrectAnswer`, not to `isSelected` alone).
 */
fun answerOptionStateFor(isSelected: Boolean, isSubmitted: Boolean, isCorrectAnswer: Boolean): AnswerOptionState = when {
    !isSubmitted && isSelected -> AnswerOptionState.SelectedUnsubmitted
    !isSubmitted -> AnswerOptionState.Default
    isCorrectAnswer -> AnswerOptionState.Correct
    isSelected -> AnswerOptionState.Incorrect
    else -> AnswerOptionState.DisabledPostSubmit
}

/**
 * `design-system/COMPONENTS.md` § Answer Options. Full-width row, height 52 (literal component-table
 * value), radius `radius.medium`, border `border.width.default`, padding `space.4` horizontal.
 *
 * **"Never color alone" (`ACCESSIBILITY.md` § 8 / the spec's own explicit rule).** [Correct]/
 * [Incorrect] each render **icon + text + color together**: [MentoraIconName.CheckCircle]/
 * `correctLabel`/`color.success.default` for Correct, [MentoraIconName.Cancel]/`incorrectLabel`/
 * `color.error.default` for Incorrect — three simultaneous, independent signals, mirroring this
 * kit's own [MentoraTextField] error-state rigor (icon+text+color together, never a bare colored
 * border). [SelectedUnsubmitted] similarly pairs its accent color with a filled check icon (the
 * spec's "radio/check filled" — reusing [MentoraIconName.CheckCircle] since no separate radio glyph
 * exists in the ported 42-icon set, rather than fabricating a new one).
 *
 * Pure presentational: [isSelected]/[isSubmitted]/[isCorrectAnswer] are plain booleans, not real quiz
 * domain models — Task 14 wires those in.
 */
@Composable
fun AnswerOption(
    text: String,
    isSelected: Boolean,
    isSubmitted: Boolean,
    isCorrectAnswer: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    correctLabel: String = "Correct",
    incorrectLabel: String = "Incorrect",
) {
    val state = answerOptionStateFor(isSelected, isSubmitted, isCorrectAnswer)
    val colorScheme = MaterialTheme.colorScheme
    val extended = MaterialTheme.extendedColors
    val opacities = MaterialTheme.stateOpacities
    val disabledText = colorScheme.onSurface.copy(alpha = opacities.disabledContent)

    val background = when (state) {
        AnswerOptionState.Default -> colorScheme.surface
        AnswerOptionState.SelectedUnsubmitted -> colorScheme.primaryContainer
        AnswerOptionState.Correct -> extended.successContainer
        AnswerOptionState.Incorrect -> colorScheme.errorContainer
        AnswerOptionState.DisabledPostSubmit -> colorScheme.surfaceVariant
    }
    val border = when (state) {
        AnswerOptionState.Default -> colorScheme.outlineVariant
        AnswerOptionState.SelectedUnsubmitted -> colorScheme.primary
        AnswerOptionState.Correct -> extended.success
        AnswerOptionState.Incorrect -> colorScheme.error
        AnswerOptionState.DisabledPostSubmit -> colorScheme.outlineVariant
    }
    val contentColor = when (state) {
        AnswerOptionState.Default -> colorScheme.onSurface
        AnswerOptionState.SelectedUnsubmitted -> colorScheme.onPrimaryContainer
        AnswerOptionState.Correct -> extended.onSuccessContainer
        AnswerOptionState.Incorrect -> colorScheme.onErrorContainer
        AnswerOptionState.DisabledPostSubmit -> disabledText
    }
    val clickable = state == AnswerOptionState.Default || state == AnswerOptionState.SelectedUnsubmitted

    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(52.dp) // component.answerOption.height (52), a literal.
            .clip(MaterialTheme.shapes.small) // radius.medium.
            .background(background)
            .border(BorderStroke(MentoraDimens.borderWidthDefault, border), MaterialTheme.shapes.small)
            .let { if (clickable) it.clickable(onClick = onClick) else it }
            .padding(horizontal = MentoraDimens.spacing.space4)
            .semantics {
                if (!clickable) disabled()
                when (state) {
                    AnswerOptionState.Correct -> stateDescription = correctLabel
                    AnswerOptionState.Incorrect -> stateDescription = incorrectLabel
                    else -> {}
                }
            },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MentoraDimens.spacing.space3),
    ) {
        when (state) {
            AnswerOptionState.SelectedUnsubmitted -> MentoraIcon(
                name = MentoraIconName.CheckCircle,
                contentDescription = null,
                size = MentoraDimens.iconSize.medium,
                tint = colorScheme.primary,
            )
            AnswerOptionState.Correct -> MentoraIcon(
                name = MentoraIconName.CheckCircle,
                contentDescription = null,
                modifier = Modifier.testTag(AnswerOptionCorrectIconTestTag),
                size = MentoraDimens.iconSize.medium,
                tint = extended.success,
            )
            AnswerOptionState.Incorrect -> MentoraIcon(
                name = MentoraIconName.Cancel,
                contentDescription = null,
                modifier = Modifier.testTag(AnswerOptionIncorrectIconTestTag),
                size = MentoraDimens.iconSize.medium,
                tint = colorScheme.error,
            )
            else -> {}
        }
        Text(
            text = text,
            style = MaterialTheme.typography.bodyMedium,
            color = contentColor,
            modifier = Modifier.weight(1f),
        )
        if (state == AnswerOptionState.Correct) {
            Text(text = correctLabel, style = MaterialTheme.typography.labelMedium, color = extended.success)
        }
        if (state == AnswerOptionState.Incorrect) {
            Text(text = incorrectLabel, style = MaterialTheme.typography.labelMedium, color = colorScheme.error)
        }
    }
}

File: mobile/iosApp/iosApp/Components/AnswerOption.swift
import SwiftUI
import shared

// Phase 5 Task T11 slice 3 (Component Kit B) -- `design-system/COMPONENTS.md § Answer Options`
// (lines 722-734). Full-width row, height 52 (literal component-table value), radius `radius.medium`,
// border `border.width.default`, padding `space.4` horizontal. Ported directly from Android's own
// `AnswerOption.kt`.
//
// **"Never color alone" (`ACCESSIBILITY.md § 8` / the spec's own explicit rule).** `.correct`/
// `.incorrect` each render **icon + text + color together**: `MentoraIconName.checkCircle`/
// `correctLabel`/`color.success.default` for Correct, `MentoraIconName.cancel`/`incorrectLabel`/
// `color.error.default` for Incorrect -- three simultaneous, independent signals, mirroring this kit's
// own `MentoraTextField` error-state rigor. `.selectedUnsubmitted` similarly pairs its accent color
// with a filled check icon (the spec's "radio/check filled" -- reusing `.checkCircle` since no separate
// radio glyph exists in this kit's 42-icon set, matching Android's own identical disclosed choice).
//
// Pure presentational: `isSelected`/`isSubmitted`/`isCorrectAnswer` are plain booleans, not real quiz
// domain models -- a later task wires those in.

/// The 5 states `design-system/COMPONENTS.md § Answer Options` defines.
enum AnswerOptionState {
    case `default`
    case selectedUnsubmitted
    case correct
    case incorrect
    case disabledPostSubmit
}

enum AnswerOptionRules {
    /// Resolves the 5-state table from plain booleans -- ported verbatim from Android's own
    /// `answerOptionStateFor`, including its exact precedence order (not submitted yet ->
    /// Default/SelectedUnsubmitted; submitted -> every option matching the real correct answer renders
    /// Correct regardless of whether the user picked it, the user's own wrong pick renders Incorrect,
    /// every other option renders DisabledPostSubmit).
    static func state(isSelected: Bool, isSubmitted: Bool, isCorrectAnswer: Bool) -> AnswerOptionState {
        if !isSubmitted && isSelected { return .selectedUnsubmitted }
        if !isSubmitted { return .`default` }
        if isCorrectAnswer { return .correct }
        if isSelected { return .incorrect }
        return .disabledPostSubmit
    }

    /// `.default`/`.selectedUnsubmitted` are the only two clickable states -- matches Android's own
    /// identical `clickable` boolean exactly.
    static func isClickable(_ state: AnswerOptionState) -> Bool {
        state == .`default` || state == .selectedUnsubmitted
    }
}

/// One (background, border, content) resolution for an `AnswerOptionState` -- pure data, directly
/// testable without rendering anything.
struct AnswerOptionColorSet: Equatable {
    let background: Color
    let border: Color
    let content: Color
}

extension AnswerOptionState {
    func colorSet(disabledTextOpacity: Double) -> AnswerOptionColorSet {
        switch self {
        case .`default`:
            return AnswerOptionColorSet(background: .mentoraSurfaceDefault, border: .mentoraBorderDefault, content: .mentoraTextPrimary)
        case .selectedUnsubmitted:
            return AnswerOptionColorSet(background: .mentoraBrandPrimaryContainer, border: .mentoraBrandPrimary, content: .mentoraBrandOnPrimaryContainer)
        case .correct:
            return AnswerOptionColorSet(background: .mentoraSuccessContainer, border: .mentoraSuccessDefault, content: .mentoraSuccessOnSuccessContainer)
        case .incorrect:
            return AnswerOptionColorSet(background: .mentoraErrorContainer, border: .mentoraErrorDefault, content: .mentoraErrorOnErrorContainer)
        case .disabledPostSubmit:
            return AnswerOptionColorSet(background: .mentoraSurfaceVariant, border: .mentoraBorderDefault, content: .mentoraTextPrimary.opacity(disabledTextOpacity))
        }
    }
}

struct AnswerOption: View {
    let text: String
    let isSelected: Bool
    let isSubmitted: Bool
    let isCorrectAnswer: Bool
    let onClick: () -> Void
    /// DISCLOSED DEVIATION from Android's own default-parameter convenience -- see
    /// `CourseProgressCard.swift`'s identical `resumeLabel` doc comment for the full rationale. Real,
    /// pre-existing `answer_option_correct_label` key.
    var correctLabel: String
    /// Same rationale as `correctLabel` above. Real, pre-existing `answer_option_incorrect_label` key.
    var incorrectLabel: String

    @Environment(\.colorScheme) private var colorScheme

    private var state: AnswerOptionState {
        AnswerOptionRules.state(isSelected: isSelected, isSubmitted: isSubmitted, isCorrectAnswer: isCorrectAnswer)
    }

    private var clickable: Bool {
        AnswerOptionRules.isClickable(state)
    }

    private var colors: AnswerOptionColorSet {
        // `MentoraButtonMetrics.StateOpacities` (`MentoraButton.swift`) does not expose
        // `disabledContentOpacity` (only hover/pressed/disabledContainer -- see that struct's own doc
        // comment), so the raw token enum is read directly here rather than widening that struct for
        // one unrelated call site.
        let disabledContentOpacity = colorScheme == .dark ? MentoraStateOpacityDark.disabledContentOpacity : MentoraStateOpacityLight.disabledContentOpacity
        return state.colorSet(disabledTextOpacity: disabledContentOpacity)
    }

    private var accessibilityStateValue: String? {
        switch state {
        case .correct: return correctLabel
        case .incorrect: return incorrectLabel
        default: return nil
        }
    }

    var body: some View {
        Button {
            onClick()
        } label: {
            HStack(spacing: MentoraSpacing.space3) {
                switch state {
                case .selectedUnsubmitted:
                    MentoraIcon(name: .checkCircle, size: MentoraIconSize.medium)
                        .foregroundStyle(Color.mentoraBrandPrimary)
                case .correct:
                    MentoraIcon(name: .checkCircle, size: MentoraIconSize.medium)
                        .foregroundStyle(Color.mentoraSuccessDefault)
                case .incorrect:
                    MentoraIcon(name: .cancel, size: MentoraIconSize.medium)
                        .foregroundStyle(Color.mentoraErrorDefault)
                default:
                    EmptyView()
                }
                Text(text)
                    .mentoraFont(.bodyMedium)
                    .foregroundStyle(colors.content)
                    .frame(maxWidth: .infinity, alignment: .leading)
                if state == .correct {
                    Text(correctLabel)
                        .mentoraFont(.labelMedium)
                        .foregroundStyle(Color.mentoraSuccessDefault)
                }
                if state == .incorrect {
                    Text(incorrectLabel)
                        .mentoraFont(.labelMedium)
                        .foregroundStyle(Color.mentoraErrorDefault)
                }
            }
            .padding(.horizontal, MentoraSpacing.space4)
            .frame(maxWidth: .infinity)
            .frame(height: 52) // component.answerOption.height (52), a literal (see file header).
            .background(colors.background)
            .clipShape(MentoraShape.medium) // radius.medium.
            .overlay {
                MentoraShape.medium.strokeBorder(colors.border, lineWidth: MentoraBorderWidth.default)
            }
        }
        .buttonStyle(.plain)
        .disabled(!clickable)
        .modifier(AnswerOptionAccessibilityValue(value: accessibilityStateValue))
    }
}

/// Applies `.accessibilityValue(_:)` only when non-`nil` -- VoiceOver's closest real analogue to
/// Android's `semantics { stateDescription = ... }` (an additional state string spoken after the
/// element's label/value), applied only for `.correct`/`.incorrect`, matching Android's own identical
/// `when` exactly (no value override for `.default`/`.selectedUnsubmitted`/`.disabledPostSubmit`).
private struct AnswerOptionAccessibilityValue: ViewModifier {
    let value: String?

    func body(content: Content) -> some View {
        if let value {
            content.accessibilityValue(value)
        } else {
            content
        }
    }
}
